internal/pkg/service/buffer/storage/local/reader/volume.go

```diff
@@ -0,0 +1,138 @@
+package reader
+
+import (
+	"context"
+	"github.com/benbjohnson/clock"
+	"github.com/gofrs/flock"
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/local"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	// lockFile ensures only one opening of the volume for reading
+	lockFile                = "reader.lock"
+	waitForVolumeIDInterval = 500 * time.Millisecond
+)
+
+type Volume struct {
+	config config
+	logger log.Logger
+	clock  clock.Clock
+	path   string
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     *sync.WaitGroup
+
+	volumeID storage.VolumeID
+	lock     *flock.Flock
+}
+
+// OpenVolume volume for writing.
+//   - It is checked that the volume path exists.
+//   - The local.VolumeIDFile is loaded.
+//   - If the local.VolumeIDFile doesn't exist, the function waits until the writer.OpenFile function will create it.
+//   - The lockFile ensures only one opening of the volume for reading.
+func OpenVolume(ctx context.Context, logger log.Logger, clock clock.Clock, path string, opts ...Option) (*Volume, error) {
```

Comment:
- Kazdy disk/volume ma svoje ID, ktore je vygenerovane a ulozene na disk ak neexistuje.
  - Umoznuje nam to lepsie riesit nejake migracie/incidenty.
  - Ak spustis `Uploader` container s nejakym diskom, tak sa `Uploader` postara o vsetky slices, ktore sa viazu k tomu disku, ... je to lepsie ako to identifikovat podla indexu/hostu v stateful sete napr. `uploader-1`.
  
Je tam aj moznost `drain` ak by sme disky potrebovali vymenit, napr. za inu stoage classe.
- Zapis novych zaznamov bude blokovany, ale upload existujucich nie.

Lock zase chrani pred nejakou misconfiguration, ak by oproti jednemu disku bezalo viacero readers/writers.